### PR TITLE
[CUDAX] Peer access control on async_memory_pool and async_memory_resource

### PR DESCRIPTION
**Pull request by pciolkosz**
This change adds member functions to enable and disable and query peer access in memory pool and async memory resource types.

It's a stateful API in CUDA that we probably need to expose as stateful API as well. There is also probably no reason to expose this through properties system, since it's tracked per device and that makes it an inherently dynamic state.

CUDA controls this state on the pool level, which means if there are two memory resources using the same pool, changing peer access on one of them will also change it on the other resource. I considered exposing these APIs only on the memory pool type, that would make it clearer about this side-effect. The problem is that the default pool will probably be the most used one and we have no way to access it outside of default constructed async memory resource. We would need to either:
1. expose some mempool_ref type returned from the resource to manipulate the pool (we might end up doing that for other reasons)
2. expose some way to get mempool object for the default pool and manipulate it instead

For now, the easiest option was to just expose this setting on the memory resource type instead, it most likely needs the peer access queries anyway, it might as well have the enable/disable APIs.

To make the peer access more convenient, this PR also adds `get_peers` member function to `device_ref`, so you can easily get a vector with all peer devices and then pass it into `enable_peer_access` on a mempool/resource.

I would like the synchronous device memory resource to follow similar APIs for the peer access, I am still exploring the options to do that.

### CODE REVIEW

File: cudax/include/cuda/experimental/__memory_resource/async_memory_pool.cuh [hunk 1]
```diff
@@ -64,6 +64,33 @@ inline void __device_supports_stream_ordered_allocations(const int __device_id)
   }
 }
 
+inline void
+__mempool_switch_peer_access(cudaMemPool_t __pool, const device_ref* __devices, size_t __count, bool __enable)
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/async_memory_pool.cuh:68]: This should be a `cuda::std:span<device_ref>` not some raw pointers

File: cudax/include/cuda/experimental/__memory_resource/async_memory_pool.cuh [hunk 2]
```diff
@@ -64,6 +64,33 @@ inline void __device_supports_stream_ordered_allocations(const int __device_id)
...
+  cudaMemAccessDesc __desc;
+  __desc.flags         = __enable ? cudaMemAccessFlagsProtReadWrite : cudaMemAccessFlagsProtNone;
+  __desc.location.type = cudaMemLocationTypeDevice;
+  for (size_t __i = 0; __i < __count; ++__i)
+  {
+    __desc.location.id = __devices[__i].get();
+    __descs.push_back(__desc);
+  }
+  _CCCL_TRY_CUDA_API(
+    ::cudaMemPoolSetAccess, "Failed to set access of a memory pool", __pool, __descs.data(), __descs.size());
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/async_memory_pool.cuh:81]: Could this rather be:
```suggestion
  const cudaMemAccessFlags __flags = __enable ? cudaMemAccessFlagsProtReadWrite : cudaMemAccessFlagsProtNone;
  ::std::vector<cudaMemAccessDesc> __descs(__count, {__flags, {0, cudaMemLocationTypeDevice}});
  ::cuda::std::tranform(__devices.begin(), __devices.end(), __desc.begin(), __set_location_id());
  _CCCL_TRY_CUDA_API(
    ::cudaMemPoolSetAccess, "Failed to set access of a memory pool", __pool, __descs.data(), __descs.size());
```

Comment (miscco) [reply cudax/include/cuda/experimental/__memory_resource/async_memory_pool.cuh:81]: Played with that and it seems worse to do it with algorithms

Comment (pciolkosz) [reply cudax/include/cuda/experimental/__memory_resource/async_memory_pool.cuh:81]: Yeah, I had it implemented with a transform first, but I ended up changing it to a simple loop, seemed a bit better

**Review by miscco — APPROVED** [review]

**Pull request merged**